crates/common/src/traces/collector.rs

```diff
@@ -0,0 +1,274 @@
+use super::shapes::Span;
+use super::resource_span_builder::ResourceSpanBuilder;
+use std::collections::{HashMap, VecDeque};
+use std::sync::Arc;
+use tokio::sync::Mutex;
+use tokio::time::{interval, Duration};
+use tracing::{debug, error, warn};
+
+/// Parse W3C traceparent header into trace_id and parent_span_id
+/// Format: "00-{trace_id}-{parent_span_id}-01"
+///
+/// Returns (trace_id, parent_span_id) as strings
+pub fn parse_traceparent(traceparent: &str) -> (String, String) {
+    let parts: Vec<&str> = traceparent.split('-').collect();
+    if parts.len() == 4 {
+        (parts[1].to_string(), parts[2].to_string())
+    } else {
+        warn!("Invalid traceparent format: {}", traceparent);
+        // Generate empty IDs if parsing fails
+        (String::new(), String::new())
+    }
+}
+
+/// Collects and batches spans, flushing them to an OTEL collector
+///
+/// Supports multiple services, with each service (e.g., "archgw(routing)", "archgw(llm)")
+/// maintaining its own span queue. Flushes all services together periodically.
+///
+/// Tracing can be enabled/disabled in two ways:
+/// 1. Via arch_config.yaml: presence of `tracing` configuration section
+/// 2. Via environment variable: `OTEL_TRACING_ENABLED=true/false`
+///
+/// When disabled, span recording and flushing are no-ops.
+pub struct TraceCollector {
+    /// Spans grouped by service name
+    /// Key: service name (e.g., "archgw(routing)", "archgw(llm)")
+    /// Value: queue of spans for that service
+    spans_by_service: Arc<Mutex<HashMap<String, VecDeque<Span>>>>,
+    flush_interval: Duration,
+    otel_url: String,
+    /// Whether tracing is enabled
+    enabled: bool,
+}
+
+impl TraceCollector {
+    /// Create a new trace collector
+    ///
+    /// # Arguments
+    /// * `enabled` - Whether tracing is enabled
+    ///   - `Some(true)` - Force enable tracing
+    ///   - `Some(false)` - Force disable tracing
+    ///   - `None` - Check `OTEL_TRACING_ENABLED` env var (defaults to true if not set)
+    ///
+    /// Other parameters are read from environment variables:
+    /// - `TRACE_FLUSH_INTERVAL_SECS` - Flush interval in seconds (default: 1)
+    /// - `OTEL_COLLECTOR_URL` - OTEL collector endpoint (default: http://localhost:9903/v1/traces)
+    pub fn new(enabled: Option<bool>) -> Self {
+        let flush_interval_secs = std::env::var("TRACE_FLUSH_INTERVAL_SECS")
+            .ok()
+            .and_then(|s| s.parse().ok())
+            .unwrap_or(1);
+
+        let otel_url = std::env::var("OTEL_COLLECTOR_URL")
+            .unwrap_or_else(|_| "http://localhost:9903/v1/traces".to_string());
```

Comment:
I already left a note at envoy template but you can directly send traces to otel collected don't need to proxy it through envoy